### DISCUSSION TITLE
Extended Library Vision

### DISCUSSION THREAD

**Discussion by nickonometry** [tyler-technologies-oss/forge-extended#16] [General]
Wanted to get a discussion started to talk about the vision and goals of the extended library. I think folks have different ideas and perceptions of what it is. Currently, the core library of Forge is very flexible and I feel like those components can handle ANY UI challenge that we throw at them. The composability of the components make them incredibly versatile which is needed in a component library built and designed for a company the size of Tyler. With that being said, we do have SOME standard UI patterns across a large majority of our apps that we really need to lock down and standardize, with minimal configuration or ability to modify. That's where the extended library can really shine in my opinion. 

This discussion all stemmed from "configuration based" vs "slots". I strongly feel that if we want these components to render a 100% consistent UI, both from a visual and functionality standpoint, we need to make them as "configuration based" as possible, meaning configuration powered by attributes and properties. Sure we may get pushback from teams trying to do something "one off", but in that scenario they have the ability to just use the core components and build their one off pattern themselves. As an example, here's two different approaches to the very simple 'confirmation-dialog' extended component:

Slot based:
```html
<forge-confirmation-dialog>
   <div slot="title">Title here</div>
   <div slot="message">Message here</div>
   <div slot="secondary-button-text">Cancel</div>
   <div slot="primary-button-text">Save</div>
</forge-confirmation-dialog>
```

Config based:
```html
<forge-confirmation-dialog 
   title="Title here" 
   message="Message here" 
   secondary-button-text="Cancel" 
   primary-button-text="Save">
</forge-confirmation-dialog>
```

Obviously there's pros and cons to each approach. But if the intended goal of the extended library is for super-enhanced consistency, then the slot based approach opens the doors for the opposite of that. If someone CAN do something weird and different, they eventually will!

And of course this component is incredibly simple, so I feel like config based works perfectly fine. Future / more complex components are definitely going to require a hybrid approach and there's no way around it

Anyways, lets discuss!
![image](https://github.com/user-attachments/assets/3d573828-bceb-4673-93f5-7b9de56b9f96)

**Comment by derekmoss** [comment]
If someone can do something weird they can and will. I'd like to remind everyone the goal of a library in general is not there to solve people problems. If they are willing and can. they will do it no matter what even if you try to restrict them. we are not in the business of trying to control people

Provide them with recipes they can copy paste. and if they do something weird we can try to evaluate whether a recipe was not clear or had a gap

It is not the job of the library to replace a lack of resourcing. It is to accelerate already capable developers.

**Comment by tyl-paul-lapczynski** [comment]
I'm typically on the side of composition over configuration. I never thought the library would be such a restricted set of components. Odds are if it's too strict, people won't even use them. Then what does the library become?

I think Derek had anything else I wanted to bring up.

**Comment by derekmoss** [comment]
The buck stops with the developer and manager, not the library. Quality assurance cannot be baked into a library. just look at accessibility. At the end of the day you can bake as much of it into a library. But whether Tyler gets sued for accessibility comes done to the developer and manager and the app they are delivering

**Reply by derekmoss** [reply]
We can accelerate a teams ability to build their app using components and recipes. but they are still responsible for them.

**Comment by Ross-Blakeney** [comment]
I think we could instead use the time that would be spent writing and maintaining the proposed config-based components to write or improve documentation to get devs to reach our expectations for the Forge design system.

It's not an unreasonable ask for devs to compose the forge-components in a manner that is consistent with the Forge design system. We should invest in our engineers instead, imo.

I don't want to discount the notion that we've tried this before and have had little success. I think that is mostly true. But we will only be creating endless technical debt for a forge-team, that does not exist, with this approach.

Is this a catch 22 since we also don't really have enough "Forge" devs available to train people? Probably. Which is why none of this will be solved without management (beyond just Ryan) acknowledging this issue.

The company needs engineers that can reliably use both the Forge component library as well as Forge the design system. There is no way around that if we want long-term success building professional applications across the board.

Composing the forge comps in a manner that is consistent with the forge design system is, at my estimation, a Jr front-end dev level task.

I think the Forge Design documentation has come a long ways, but that is prob the weakest link. Mainly, I think teams need to see more examples of what the design system's expectation is for all patterns. That could be more of an issue with teams not even looking at the documentation. Is it a discoverablity issue? I'm not sure. But I do know that many teams are using napkin mocks from their PO that may be sloppily slapping buttons on the wrong side of a dialog, etc.. just because they are trying to move fast.

I'd rather we see a push to get people using the existing docs and see where we can improve on content and discoverablity. Even accountability at both the business and dev level. But, again, that requires management buy-in.

**Comment by DRiFTy17** [comment]
I am in agreement with everything mentioned above, and I'd like to throw in an additional restriction (aside from just composition) that comes along with providing content via attribute vs slots. Text provided via an attribute will always be just plain text (unless we unsafely/sanitize and render HTML content which is a no-go). This means that there can never be text decoration such as bolded portions, italics, line breaks or any other adjustments. Which IMO is too restrictive for arbitrary content, and it's not the job of the UI component to _attempt_ to police these design deviations.

The goal of the extended library IMO is to drastically improve developer velocity by taking on the work to create common design patterns, and allow developers to copy/paste into their app, adjust some content, and move on to the next thing. The benefit of these components will be in the layout and placement of content for visual consistency, not policing the developer-provided content that can be rendered within. Just my two cents.

**Reply by MikeMatusz** [reply]
I think this is an important consideration.  Using the confirmation dialog as an example, could the message contain a hyperlink for more info (maybe bad practice, just throwing out ideas), or include a list of records that would be affected by the operation (we do this frequently, though mostly warnings, not confirmation)?

I do think a hybrid approach could be called for.  Sometimes there are parts of a UI that are very deliberately being abstracted for consistency/accessibility/behavior, and in those cases I'd prefer to keep it configuration-based.  The PSDF has a number of components that have several content slots, but then specific inputs for things we might need to use in multiple place or have to intentionally constrain.  Like maybe we need label as a string, so that becomes an input rather than a slot.

It also is a lot more verbose to do `<div slot="primary-button-text">Save</div>`, and then should I use a div or span?   Configuration is simpler there.   But if we're packaging up a confirm dialog, are we also packaging up similar info/warning/error dialogs?  What if I need a "confirm with reason" dialog with a textarea to enter a reason for deleting this record for auditing.  Or could there be a scenario where I want a third button?

In Public Safety, we created two level of abstraction.  One was a generic composition based "message-box-dialog" component, and then provided configuration based ways to show standard confirm, error, and warning dialogs that leverage it.  But for those edge cases, like confirm with reason, we can still leverage the generic component to easily create an experience consistent with the standard dialogs.  Not saying everything config-based needs to be backed by a composition-based component, but if we know from the start there's a need for multiple flavors, we don't need to assume our list of those is exhaustive.

**Reply by samrichardsontylertech** [reply]
I've been questioning the approach of even having a dedicated confirmation dialog. What's the difference between this, a warning dialog, information dialog, etc? Just the format of the message? They all have a title, simple body content, and one or two buttons. Are we going to have a component for each slight permutation? Why not have one that has sensible defaults but can be adapted with minimal configuration?

This is kind of a conceptual question, but we also run the risk of contributing to bloat and excessive maintenance if we go too far down this path.

**Reply by samrichardsontylertech** [reply]
Also, I'm in support of having a configuration option in addition to slots, mostly because we usually show dialogs programmatically

**Reply by nickonometry** [reply]
I don't see any harm in having multiple variations of similar components, ie: confirmation dialog & a warning dialog & informational dialog. Or altering the confirmation dialog to have some sort of theme attribute that renders slightly different opinionated UI (thinking out loud here, but this could include some sort of baked in icon/illustration, red text, etc), while leaving the important aspects of it in tact, ie: the title and message. 

And I also wanted to add to Mikes comment above, this confirmation dialog is a good example of where a simple string makes sense for the message element, based on the design guidance and use case of the component. We don't want any bolding, italicized, text, hyperlink, lists etc, it's literally used for a couple informative sentences about a destructive action that occurred with a primary and optional secondary action. In that very specific use case where we are confident in what we need, slots are overkill in my opinion

**Comment by samrichardsontylertech** [comment]
Ideally I'd like to keep our design system and component libraries separate. We should be able to trust designers to follow our guidelines and developers to implement the designs faithfully.

I actually think an overly strict approach could hurt consistency. If a design calls for something like this confirmation dialog but requires a slight change, maybe including a list of documents or errors, the developer will be forced to almost-recreate this from the ground up and probably introduce way more errors that also break accessibility. Never mind that there are already endless ways to break a design that aren't enforceable at a component level — capitalization, wording, etc. A reasonably flexible component backed by firm design guidance is a stronger way to go, IMO.

From a dev perspective, I just think the config-based approach is unergonomic and confusing. I'm uneasy with content being set via an attribute.

**Comment by IdesOfMarchio** [comment]
To my understanding the dividing line on configuration vs composition of Forge is around visual consistency and developer convenance.

If you look at industry leading libraries like IBM Carbon or MUI, they utilize both a flavor of config and composition. Some leaning more to one side than the other so do we really need to pick a side?

We've seen MAJOR value in a configuration leaning library because its easier to toss a "Full Stack" developer to use a config leaning library because the api is more straightforward and documentation requires less of a learning curve. That alone was a huge win for us because the devs didn't need to take too much time understanding the library in detail and instead could use the components as designed. We do have 'slots' and other composition based code in some instances though and utilize TS to enforce the proper usage of those slots. In example, we don't want to allow devs to slot a table into an omnibar 'because the client asked for it' or more realistically embed some complex widget into a table because they 'want to see all the data and more'. 

The restriction of configuration ultimately enables speed and forces visual consistency by not allowing devs to make changes that are not needed or within the design system at the 'burden' of developer control. 

I've heard "we need more trained FEDs" for years now... gunna be a realist here and say that isn't happening in most shops and would rather focus on designing a library that is easier to use instead of a library that gives the advanced dev the full control they are looking for.

**Reply by IdesOfMarchio** [reply]
tl;dr
It's easier to train someone to use a wrench to fasten a bolt than train someone to weld two pieces together. They both hold items together, and the welding gives the user more control of how the items are put together, but from metrics, most clients just need the items bolted together.

**Reply by MikeMatusz** [reply]
I agree with a lot of this, but I definitely advocate hybrid over 100% configuration.  There's always going to be some assumptions made that won't hold up, and then you end up with either extra complexity, extra components, or teams not using it and making something on their own which doesn't meet the goals of forge-extended enforcing consistency or being an accelerator. 

Config-based can be simpler to use, and sometimes it's necessary to enforce some restrictions for a component to function properly or provide additional value, and those are both good arguments for it.  Simpler isn't better if it falls short too often, though.  If it feels like a burden to provide a slot for button text, then maybe it is a good candidate for abstracting, but I think generally a lot of shared components need some flexibility.  If you really need to simplify it further, you can always wrap another layer around it that's 100% config-based, but I'd be hesitant to make that the only option for a shared component meant to be used across all the diverse contexts within Tyler (plus potentially any consumers of it as open source?).  Keep it as simple as it should be, but no simpler.  Probably need more of a decision tree than that, but I think some kind of balance will serve best.

**Reply by nickonometry** [reply]
Agreed Mike! In the scenario of the confirmation dialog and it's very specific use case, 100% configuration makes the most sense in my opinion. I don't forsee any future extended components being 100% configuration based, it would have to be something extremely simple that we're not thinking of. Everything else will almost certainly have to be a hybrid approach

**Reply by DRiFTy17** [reply]
I think the confirmation dialog _could_ be a hybrid. The only slot available could be for the message and the rest attributes (configuration), just to support the cases where some other more rich content needs to be used in the message area but we don't want to make devs go recreate the whole thing just to get around that restraint. It can also support both a message attribute and a slot to get the best of both worlds, but the default docs recommend configuration.

**Comment by samrichardsontylertech** [comment]
At some point we also need to clarify what the purpose of the extended library is. Is it a repo for pre-implemented recipes 1:1 with the design system? Or is it an extension of the core component library with more complex and niche components?

**Reply by samrichardsontylertech** [reply]
Building on this, is the extended library intended only for internal usage or is it for public use like the core library? If it is public, does it make sense to tie it so strongly to our internal guidance?

**Reply by IdesOfMarchio** [reply]
That's a fair question, I thought the intend was internal usage

**Reply by samrichardsontylertech** [reply]
My understanding was that this would be more like [Angular Material CDK](https://material.angular.io/cdk/categories) (in a sense) and built-out recipes would be part of Forge Awesome. I don't think we worked this out enough before committing to it.

**Reply by MikeMatusz** [reply]
The forge-extended vs forge-awesome distinction is definitely unclear.  I originally thought forge-extended was the implementation of the "forge-awesome" concept that was tossed around, but now it seems like those are actually two separate things.

**Reply by DRiFTy17** [reply]
For what it's worth, we haven't actually released anything yet so this can still be determined. I'll wait to get consensus before publishing any packages.

**Reply by DRiFTy17** [reply]
"awesome-forge" = curated collection of forge-related items. Basically anything Forge-related that can be shared with others whether that is dev-related or not.

"forge-extended" = A library of larger prebuilt/composite UI patterns as reusable components. This would be a link in the curated "awesome-forge" repo essentially.

"recipes" = code examples/demos of using multiple Forge components (both atomic and extended among other things) to create a larger design pattern as a starting point with the intention to be potentially customized to fit the needs of the use case.

**Comment by ryansheehan** [comment]
I am late to this party.  But I want to reiterate the high level goal of Forge, because I think it applies directly to this conversation.

## Objective of Forge

> Forge solves the need for a framework independent toolkit that provides the flexibility for the diverse set of applications of Tyler while providing a quintessential cohesive Tyler look-and-feel.

The design considerations for a _back office court clerk_ is not the same as a _policemen on a stakeout_ is not the same as a _school bus driver_ or a _fire inspector in a restaurant_.

## Nature of developers

As developers, we solve problems.  Typically, if our frameworks do not directly address our problems, we find another solution.  Most of us do not raise our hand and stomp our feet asking for someone else to clear the hurdle.  So what happens if Forge does not provide the developer the path they need?

## The slippery slope

If we lock down components through configuration, we may improve DX, but we remove autonomy from engineers.  They will solve the problem in another way, likely outside of Forge.  If enough of these cases appear, will Forge be selected for their next project? I think the perceived lack of robustness would dissuade its selection.  Our experience with TCW (aka Forge v1) informed us that developers will seek alternate solutions.  We will fail to live up to high level goal of Forge if we are too prescriptive in its api.

### The example

Let's look at @nickonometry straw man above.  It solves for the need for a _title_ on the dialog.  Given both potential interfaces, what happens when a requirement comes along that a dialog requires a _subtitle_?

The solution is easy within the composition pattern:

```html
<forge-confirmation-dialog>
   <!-- <div slot="title">Title here</div> -->
   <hgroup slot="title">
       <h2>Title Here</h2>
       <p>Subtitle Here</p>
   </hgroup>
   <div slot="message">Message here</div>
   <div slot="secondary-button-text">Cancel</div>
   <div slot="primary-button-text">Save</div>
</forge-confirmation-dialog>

/*
After-thought here: semantic element choice can be an issue.  
In composition, the semantic element choice is left to the consumer.  
In a configuration setup, we need to remember to expose an option. 
In this straw man example, how many semantic options should we expose?  
Should it be button text, link text, and would we need an additional 
attribute to set the appropriate anchor/button element?
*/
```

In a configuration pattern, the likely path is that the engineer won't even complain, but just build out their own custom solution.  The problem will get solved, the solution will not be shared, and this dialog variant will be re-solved again by another team in the future. At best the problem is raised to our governance, but we would still have to extend the existing configuration based component and release a new version to support the variant.

## Composition draw backs

The composition pattern is more verbose.  There is no way around this, it's the language of HTML.  But its direct access to HTML that makes it so appealing.  The other main drawback, that I see, to composition, is that it's more in the face of engineering management when a front end skill gap is present on their team.  You need engineers who understand the tools being used.

## Configuration vs composition

Forge has never been, nor ever will be, designed to address front-end skill gaps.  There are many nuances in front end development that could never be solved for through the usage of Forge. A composition system allows for a larger surface area for front end engineers to use the tools of HTML/CSS/JS to address nuanced gaps in the library. A configuration based api, by design, will not be as robust to handle unforeseen gaps in the framework.

## Vision for Forge Extended (and other Forge-adjacent concepts)

My personal vision around Forge Extended is a library of components of which they themselves are composed from the core componentry.  Forge Extended components have **low to no** design volatility across products.  If a design pattern has high design volatility then they should be classified as a _recipe_.  If a component is built around high design volatility then it should be shared/marketed through the Forge Awesome collection.  Forge Extended is intended to be maintained by the maintainers of Forge Core.  Forge Awesome is intended to be linked out to other repositories and those components are maintained by the submitter.  Forge Recipes are snippets of code designed to be a starting block for engineers designed with intent for heavy modification.

**Reply by nickonometry** [reply]
I definitely agree with all of the above @ryansheehan, especially the low to no volatility statement in the past paragraph. The problem I see with completely composable extended components is that the volatility can be endless. My only goal with this particular component (the confirmation dialog) is to achieve the exact design specs that we recommend in Forge (which will be adjusted to match this extended component based on all the variants we've collected and seen). That means this needs an optional title, 1 or 2 paragraphs in the body with no formatting (bolds, italics, etc. or hyperlinks), an optional secondary button, and a primary button. If a team needs something different they could easily compose it however they need. But for 99% of the confirmation dialogs out there, that is all that's needed. 

Overall the end goal is to achieve this:
![image](https://github.com/user-attachments/assets/7c813fe2-270a-43b3-8b5a-56421eeacd77)

And prevent this:
![image](https://github.com/user-attachments/assets/f04fd601-ff2d-4b54-aaf0-76c227c4ace4)
